### PR TITLE
Add django-tables2 and utils, example, docs for usage with HTMX

### PR DESCRIPTION
**Pull request by biyeun**
## Technical Summary
This PR introduces a new python dependency `django-tables2`, which we will use to paginate and sort tabular data. In combination with HTMX, this library allows us to quickly render partial template responses of a table's page based on a `queryset` and a `requests` `GET` parameters. In combination with `HTMX`, this introduces a powerful way to easily asynchronously paginate through tabular data.

To support this combination of libraries, this PR introduces:
- custom `django-tables2` templates for Bootstrap 5 tables and Bootstrap 5 tables with `HTMX`
- A `SelectablePaginatedTableView` base class based on `django-tables2` `SingleTableView` which handles the widget selecting page length and saving that selection to a cookie (and fetching the saved value later)—similar to how our Knockout pagination component handles selecting pagination length.
- New `scss` styles for sortable headers in a `django-tables2` table
- A new custom `hq-hx` plugin called `hq-hx-loading` for triggering a loading overlay (in addition to the usual `hx-indicator` functionality...you can't easily do both at once with `hx-indicator`). We use this in our `django-tables2` `HTMX` table.
- A new `hq-hx-refresh` plugin for "chaining" an `hqRefresh` event to a target selector (the value of the `hq-hx-refresh` attribute) once the `HTMX` request on the originating element with that attribute completes. (`hq-hx-refresh-after` for `htmx:afterRequest` and `hq-hx-refresh-swap` for `htmx:afterSwap`).
- A `BaseHTmxTable` class to make setting up the `django-tables2` table definition object easier.
- A working example added to the styleguide using all the features described above.
- Documentation updates

Screenshot 2024-12-03 at 9 19 33 PM
Screenshot 2024-12-03 at 9 20 20 PM
Screenshot 2024-12-03 at 9 20 00 PM
Screenshot 2024-12-03 at 9 19 45 PM

### Safety story
Safe change. Documentation updates and utilities. No changes to active pages.

### Automated test coverage
No

### QA Plan
Not needed

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/hqwebapp/templates/hqwebapp/tables/bootstrap5_htmx.html [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    </div>
+  </div>
+{% endblock %}
+
+{% block select-per-page-attr %}
+  name="{{ table.per_page_field }}"
+  hx-get="{{ request.path_info }}"
+  hx-target="{% if table.css_id %}#{{ table.css_id }}{% else %}div.table-container{% endif %}"
+  hx-swap="outerHTML"
+  hq-hx-loading="{{ table.loading_indicator_id  }}"
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/hqwebapp/templates/hqwebapp/tables/bootstrap5_htmx.html]: I like the shortened name, but I feel it abstracts the `hx-` (for `HTMX`) connection a little too much. It's clear from `hq-hx-` that this is a custom HQ HTMX attribute. And since Alpine attributes begin with `x-`, `hqx-` seems more ambiguous as it could reference alpine, not `HTMX`. I would prefer to keep the naming scheme as is. Regardless, I feel such a change would be out of scope for this PR given that there are other extensions existing that already a part of HQ.

File: corehq/apps/hqwebapp/static/hqwebapp/js/htmx_utils/hq_hx_loading.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+/*
+    Adds an `is-loading` class to the element with ID specified in the `hq-hx-loading` attribute.
+    This `is-loading` class is applied when to that ID before an HTMX request begins, and is
+    removed after an HTMX swap is completed.
+
+    This is useful for adding loading indicators to elements outside the parent heirarchy available
+    through using `hx-indicator` alone. Right now, this is used to add an `is-loading` style to a django tables
+    table, which overlays a loading indicator across the entire table (seen in hqwebapp/tables/bootstrap5_htmx.html)
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/hqwebapp/static/hqwebapp/js/htmx_utils/hq_hx_loading.js:8]: it's important for paginating the table that this workflow is in place, as clicking on an edit button, select row, etc during pagination would result in issues. Similarly during the whole table refresh process when filters are applied. I think if you see the data cleaning example on staging https://staging.commcarehq.org/prototype/dc/ (you need the FF enabled), then that will give you the interactive example. 

TL;DR Cell-specific actions are already limited. The whole-table barrier is intentional and necessary.

I believe an error would still fire that event. Regardless, if there is an error on pagination, I think we have bigger issues...I would want a disruptive behavior. In any case, refresh is always possible.

File: corehq/apps/hqwebapp/tables/htmx.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        if not hasattr(self, 'container_id'):
+            self.container_id = self.__class__.__name__
+        if not hasattr(self, 'loading_indicator_id'):
+            self.loading_indicator_id = f"{self.container_id}-loading"
+
+    class Meta:
+        attrs = {
+            'class': 'table table-striped',
+        }
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/hqwebapp/tables/htmx.py:49]: I really prefer the other way tbh

File: corehq/apps/styleguide/utils.py [hunk 4]
```diff
@@ -1,35 +1,20 @@
...
-             random.choice(colors),
-             random.choice(big_cats),
-             months_from_date(today, -1 * random.choice([6, 12, 24, 48])).strftime("%Y-%m-%d"),
-             random.choice(apps),
-             months_from_date(today, -1 * random.choice([0, 1, 2, 3])).strftime("%Y-%m-%d"),
-             random.choice(case_owners), random.choice(status)]
-        )
+        rows.append([
+            "patient",
+            fake_data.get_first_name() + fake_data.get_last_name(),
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/utils.py:11]: whoops. thanks!

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    )
+    date_opened = columns.Column(
+        verbose_name=gettext_lazy("Opened On"),
+    )
+    owner = columns.Column(
+        verbose_name=gettext_lazy("Owner"),
+    )
+    status = columns.Column(
+        verbose_name=gettext_lazy("Status"),
+    )
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py:45]: eh...I find the other format more readable tbh

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    The variable names for each column match the keys available in
+    `generate_example_pagination_data` below, and the `verbose_name` specifies
+    the name shown to the user.
+
+    We are using the `BaseHtmxTable` parent class and defining its Meta class
+    below based on `BaseHtmxTable.Meta`, as it provides some shortcuts
+    and default styling for our use of django-tables2 with HTMX.
+    """
+    class Meta(BaseHtmxTable.Meta):
+        pass
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py:20]: no it is absolutely necessary. it is inheriting from `BaseHtmxTable.Meta`.

```
class Foo(BaseHtmxTable):
    ...
```
this by itself is not enough if `BaseHtmxTable` has `class Meta` inside of it. Believe me, I tried. This is the way.

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    pagination of a given queryset based on GET parameters in the request.
+
+    This view will be fetched by the "host" `HtmxPaginationView`
+    via an HTMX GET request.
+    """
+    urlname = "styleguide_b5_paginated_table_view"
+    table_class = ExampleFakeDataTable
+
+    def get_queryset(self):
+        return generate_example_pagination_data(100)
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py:19]: I have not yet. I'm investigating that after I return from PTO in January. This is intentionally simple as it's just an example. It will likely involve a custom paginator class, I'm sure of it...especially with elasticsearch data.

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+from corehq.apps.hqwebapp.tables.pagination import SelectablePaginatedTableView
+from corehq.apps.styleguide.examples.bootstrap5.htmx_pagination_data import generate_example_pagination_data
+from corehq.apps.styleguide.examples.bootstrap5.htmx_pagination_table import ExampleFakeDataTable
+
+
+class ExamplePaginatedTableView(SelectablePaginatedTableView):
+    """
+    This view returns a partial template of a table, along with its
+    page controls and page size selection. Its parent classes handle
+    pagination of a given queryset based on GET parameters in the request.
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py:10]: that's part of the parent classes this class inherits from that's built into the `django-tables2` library.

The pages themselves are handled in the template.

see https://github.com/dimagi/commcare-hq/blob/0ca8a1d96de0a16d4a1ddbf9fa5f447a6d53f3bc/corehq/apps/hqwebapp/templates/hqwebapp/tables/bootstrap5_htmx.html#L34-L78

and
https://github.com/dimagi/commcare-hq/blob/ce5dc407c5d67c2b9f545c56a594d9e79ade9ce8/corehq/apps/hqwebapp/templates/hqwebapp/tables/bootstrap5.html#L54-L109

for usage within our templates.

File: corehq/apps/hqwebapp/tables/htmx.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+class BaseHtmxTable(tables.Table):
+    """
+    This class provides a starting point for using HTMX with Django Tables.
+
+    usage:
+
+        class MyNewTable(BaseHtmxTable):
+
+            class Meta(BaseHtmxTable.Meta):
+                pass  # or you can override or add table attributes here -- see Django Tables docs
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/hqwebapp/tables/htmx.py:15]: `pass` is actually intentional. It's always required, unless overrides or additions are happening.

File: corehq/apps/hqwebapp/templatetags/hq_tables_tags.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    """
+    Based on
+    https://stackoverflow.com/questions/31838533/django-table2-multi-column-sorting-ui/31865765#31865765
+    For allowing multiple column sorting
+    """
+    context['use_htmx_links'] = link_type == 'htmx'
+    sorted_columns = context['request'].GET.getlist('sort')
+    for column in context['table'].columns:
+        column.sort_existing = column.name in sorted_columns
+        column.sort_done = f"-{column.name}" in sorted_columns
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/hqwebapp/templatetags/hq_tables_tags.py:17]: I was attempting multi-column sorting at one point, and missed refactoring this. I'll just simplify it now to reduce the confusion

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py:20]: That's unfortunate and not how Django model classes work. It would be possible to add an empty/default `Meta` class to any `BaseHtmxTable` that does not define it. Might have to do it with a "metaclass" (not to be confused with the `Meta` class).

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py:45]: I guess we can agree to disagree. I find the compact style MUCH more readable.

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py:10]: I think the original question was poorly worded. I was not asking how the templates use HTMX to add GET parameters to the request. I was asking where is the code that retrieves GET parameters from the request to construct/filter the database query? Is that code easy to extend or is it deep in the framework somewhere where where it is difficult to customize?

Based on [this comment](https://github.com/dimagi/commcare-hq/pull/35470/files#r1889239620) it sounds like you will be looking into this more later. It feels like an important design detail in this PR, so it doesn't feel great to sign off on this without understanding more about how that will work.

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table.py:20]: That creates some complications as there is already a metaclass for a `django-tables2` table that references the attributes defined in `Meta`. I spent some time trying to figure out a workaround, but this was the simplest way that didn't add additional abstractions to the expected behavior from `django-tables2` At least the same attributes don't have to be re-writted each time the same kind of htmx table is used

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py:10]: https://github.com/dimagi/commcare-hq/blob/24853e4ce8afd1eddd20bfdab47d7b1f275ac7c4/corehq/apps/hqwebapp/tables/pagination.py#L42-L46

`SelectablePaginatedTableView` subclasses `ListView`, which is a built-in `django` view that handles the pagination based on a given paginator and data source. the table mixin provides the tabular data source and structure for the template tags

[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py:10]: yes, digging into the code, the pagination happens from django's `Paginator`. `django-tables2` formats the data for display, but ultimately it's django's core classes that are doing the work, and they seem extendable.

**Review by millerdev — APPROVED** [review]
Met with @biyeun and confirmed that more backend logic, including a new paginator class, implementing efficient deep pagination will be coming in a follow-up PR.

File: corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/styleguide/examples/bootstrap5/htmx_pagination_table_view.py:10]: FYI @millerdev here is the followup PR to this question https://github.com/dimagi/commcare-hq/pull/35712
